[RHDEVDOCS-6657] Release notes for Pipelines 1.20 (#98437)
openshift/openshift-docs: modules/op-release-notes-1-20-0.adoc

.{tekton-results}
* {tekton-results} supports PostgreSQL version 17, ensuring compatibility for Konflux transition from version 13. This upgrade enables a smooth migration of Konflux's Relational Database Service (RDS) instance of the database from version 13 to either version 16 or 17. 

* {tekton-results} skips processing of already stored `PipelineRuns`, improving performance and reducing API server calls. This optimization enhances overall efficiency and responsiveness, aiming to reduce server load and conserve resources.

* {tekton-results} supports a new flag for the `tekton-results-watcher` controller to disable live collection of Runs by setting the following flag in the `TektonConfig` CRD:
+
[source, yaml]
----
#...
options:
    deployments:
      tekton-results-watcher:
        spec:
          template:
            spec:
              containers:
              - args:
                - "--disable_storing_incomplete_runs=true"
                name: watcher
                resources: {}
#...
----
+

> If set to `true`, runs are not stored until they are complete. If `false`, runs are stored upon creation and continuously upserted until deletion (current default). Using `true` helps improve performance.

Reviewer: continuously upserted whenever there's a change.

Probably, we can mention that this will become the default behaviour in coming releases. So you might need to explicitly set it to false.

* All containers in the `tekton-results` deployment have `readOnlyRootFilesystem` setting enabled, improving security and compliance.

.{pac}
* {pac} synchronizes the status of running `PipelineRun` on git providers when managed by `Kueue`, a Kubernetes job-queuing system. This enhancement helps developers receive feedback on their pull requests faster through comments and status checks.

> * {pac} supports passing webhook parameters in the JSON POST body, while maintaining compatibility with URL query parameters. Using the request body for sensitive information reduces exposure in logs and improves security.

Reviewer 2: @ochromy I think it should contain example. I don't know how to format it you can a look at that.

* {pac} supports detailed logging for GitHub API calls, providing insights into API interactions, durations, and rate-limiting. By setting the controller's log level to 'debug', you can troubleshoot complex issues more efficiently. This enhancement refactors code to instrument GitHub calls and affects all types of GitHub API calls within the provider.

> * {pac} adds new `auto-configure-new-github-repo` setting in to consolidate repository creation within a single namespace, streamlining management for those with multiple repositories. This helps in creating more organized and manageable environment by autoconfiguring repository names, eliminating the potential chaos caused by multiple namespaces for each repository. When using this new setting, you can also configure your custom repository name by using the `auto-configure-repo-repository-template` setting.

Reviewer 2: ```suggestion
* {pac} adds new option in `auto-configure-new-github-repo` feature setting to consolidate repository creation within a single namespace, streamlining management for those with multiple repositories. This helps in creating more organized and manageable environment by autoconfiguring repository names, eliminating the potential chaos caused by multiple namespaces for each repository. Now, you can configure your custom repository name by using the `auto-configure-repo-repository-template` setting.
 
* see example:
 
* auto-configure-new-github-repo: "true"
* auto-configure-repo-namespace-template: "all-repositories-ns"
* auto-configure-repo-repository-template: "{{repo_owner}}-{{repo_name}}-cr"

when you configure these settings, {pac} will create Repository CR for all the newly created GitHub repositories in `all-repositories-ns` namespace. for exmaple if you create a repository on GitHub with name `org/test-repo` a Repository CR will be created in `all-repositories-ns` namespace like this `org-test-repo-cr`.
```

* You can use relative paths to reference tasks from within a remote `Pipeline` definition. The {pac} resolver automatically builds the full URL for the task based on the location of the remote `PipelineSpec` definition. This improves pipeline organization and portability by keeping tasks and pipelines co-located, which is especially useful when working with tagged releases or branches.

.Helm
* The Operator Helm chart supports init containers. When you set `operator.initContainers.enabled` setting to `true`, you can configure and execute custom init containers before the Operator starts. You can use this to inject pre-start tasks or configurations.

[id="pipelines-breaking-changes-1-20-0_{context}"]
=== Breaking changes

> * {pac} automatically moves from using the deprecated {tekton-hub} to {artifact-hub} for task resolution. After the migration some task references using short version pins, such as `0.2`, will not resolve correctly. You must update these pins to the full semantic version format, such as `0.2.0`, to ensure correct task resolution.

Reviewer 2: ```suggestion
* {pac} automatically moves from using the deprecated {tekton-hub} to {artifact-hub} for pipeline, task resolution. After the migration some task references using short version pins, such as `0.2`, will not resolve correctly. You must update these pins to the full semantic version format, such as `0.2.0`, to ensure correct task resolution.
```

* Before this update, git resolver deployments could leave zombie git processes running, consuming resources. With this update, resolvers  use a `tini`-based image and entrypoint to actively clean up subprocesses.

* Before this update, upgrading Tekton could prevent updates to metadata, such as finalizers, on a completed `PipelineRun` or `TaskRun` due to specification drift. With this update, you can  update metadata on completed PipelineRuns and TaskRuns immediately after upgrades.

* Before this update, the OpenShift git resolver did not mount the trusted CA ConfigMap into the component system CA store, potentially causing certificate verification issues. With this update, the ConfigMap is mounted correctly, ensuring secure git operations.

* Before this update, the `git-clone` task failed with a `No such remote 'origin'` error messgae if the `origin` remote was missing from the repository. With this update, the task automatically adds the `origin` remote to the repository configuration, ensuring correct setup and successful cloning.

> * Before this update, streaming live logs for a PipelineRun or TaskRun with the {tekton-results} CLI required the `-f` flag. Attempts to query logs without the flag resulted in errors, and status messages for running resources were unclear. With this update, the CLI streams logs without requiring the `-f` flag, and provides clearer status messages when querying logs for running `PipelineRuns` and `TaskRuns`.

Reviewer 3: For some reason, this is coming inside the "pipelines" section. Which is not correct.
Also, this point is conveying the wrong information. Correct point should be:
```
Before this update, users querying logs for a running PipelineRun or TaskRun in OPC Results received an unclear error message. With this update, the message clearly indicates that logs cannot be retrieved while the run is still in progress.
```

.Pruner
* Before this update, the event-based pruner failed to process `PipelineRun` objects when the `historyLimit` setting was set to `0`, leaving them in a running state. With this update, the pruner handles `PipelineRuns` with `historyLimit` set to `0`, improving resource management.  

* Before this update, the pruner could panic during upgrades if a setting was `nil`. With this update, the pruner handles `nil` values correctly and avoids crashes.

.{pac}
* Before this update, structured logs for {pac} lacked detailed source repository information from the initiating webhook request. With this update, logs for {pac} include complete source repository details, making it easier for operators to identify user-reported issues.

> * Before this update, {pac} failed immediately when resource quotas were exceeded, canceling the run and interrupting user workflows. With this update, {pac} controllers retry and automatically rerun if resources become available, reducing unnecessary cancellations and improving pipeline reliability.

Reviewer 2: @ochromy thanks! actually it is not PaC bug it was `tektoncd/pipeline` component bug so you should move it to there

* {tekton-results} skips processing of already stored `PipelineRuns`, improving performance and reducing API server calls. This optimization enhances overall efficiency and responsiveness, aiming to reduce server load and conserve resources.

* {tekton-results} supports a new flag for the `tekton-results-watcher` controller to disable live collection of Runs by setting the following flag in the `TektonConfig` CRD:
+
[source, yaml]
----
#...
options:
    deployments:
      tekton-results-watcher:
        spec:
          template:
            spec:
              containers:
              - args:
                - "--disable_storing_incomplete_runs=true"
                name: watcher
                resources: {}
#...
----
+
If set to `true`, runs are not stored until they are complete, improving system performance. If set to `false`, runs are stored upon creation and continuously upserted whenever a change occurs. This is the current default behavior. The `true` setting is planned to become the default in a future release.

* All containers in the `tekton-results` deployment have `readOnlyRootFilesystem` setting enabled, improving security and compliance.

.{pac}

> * {pac} synchronizes the status of running `PipelineRun` on git providers when managed by `Kueue`, a Kubernetes job-queuing system. This enhancement helps developers receive feedback on their pull requests faster through comments and status checks.

Reviewer 2: @anithapriyanatarajan thank you for bringing this in attention, Yeah this is only for Konflux and I think should not be gone into release notes. @ochromy can you please remove this?

* {pipelines-shortname} adds anti-affinity rules to High Availability (HA) setups, ensuring that replicas are not actively deployed on the same node. This improves system resilience and availability by distributing replicas across multiple nodes, reducing the risk of a single node failure affecting multiple replicas. This improves service availability and resilience, better load balancing, and overall system performance.

* * With this update, you can independently control whether the {pipelines-shortname} Operator creates Role-Based Access Control (RBAC) resources and Trusted Certificate Authority (CA) bundle ConfigMaps in your cluster. Two new options in the `TektonConfig` custom resource (CR) allow you to configure this behavior:  
+

> ** `createRbacResource`: Enables or disables creation of RBAC resources.  

Reviewer 4: for new features, we need to specify under particular component similar to 1.19. For example - this should be under operator section

* Pull requests that trigger a push event populate a dynamic `pull_request_number` variable automatically, allowing for a clear reference to the specific pull request associated with the push event and improving traceability.

* {pipelines-shortname} Operator references the up-to-date task image, ensuring the correct image version is used for pipeline tasks.

.Pruner
* You can enable and configuer the event-based `tektonpruner` directly in the `TektonConfig` custom resource (CR).

* The event-based pruner includes a telemetry service `ServiceMonitor` that tracks pruned `PipelineRuns` and `TaskRuns` on port `9090`. Metrics support monitoring, alerting, and debugging. This enables Prometheus-based monitoring of pruner metrics, improves observability, and requires no additional configuration.

.{tekton-triggers}
* You can choose whether to install {tekton-triggers} when deploying {pipelines-shortname} through the Operator. This enhancement provides finer control over component installation, allowing flexibility for environments where triggers are managed separately.

.{tekton-results}

> * {tekton-results} supports PostgreSQL version 17.

Reviewer 4: please remove {tekton-results} supports PostgreSQL version 17. - its not needed

You can also configure custom repository names with the `auto-configure-repo-repository-template` setting.
+
[source, yaml]
----
#...
    auto-configure-new-github-repo: "true"
    auto-configure-repo-namespace-template: "all-repositories-ns"
    auto-configure-repo-repository-template: "{{repo_owner}}-{{repo_name}}-cr"
#...
----
+
With this example configuration, {pac} creates a `Repository` custom resource (CR) for every new GitHub repository in the `all-repositories-ns` namespace.  
For example, if you create a repository named `org/test-repo` in GitHub, {pac} creates the corresponding `Repository` CR in the `all-repositories-ns` namespace as `org-test-repo-cr`.

* You can use relative paths to reference tasks from within a remote `Pipeline` definition. The {pac} resolver automatically builds the full URL for the task based on the location of the remote `PipelineSpec` definition. This improves pipeline organization and portability by keeping tasks and pipelines co-located, which is especially useful when working with tagged releases or branches.

> .Helm

Reviewer 4: we shoudnt have section for helm, it should be included in the component section, for this its on operator

* Before this update, running `fbc-fips-check-oci-ta` taks during git cloning caused temporary resource unavailability, stalling the git resolver and preventing the resolution of `ResolutionRequests`. This resulted in a build failure with the following error message: `cannot fork() for remote-https: Resource temporarily unavailable`. With this update, the issue is fixed.

* Before this update, `TaskRun` and `CustomRun` creation could fail immediately on transient mutating-webhook timeouts, causing flakiness on busy clusters. With this update, the system  applies exponential backoff, configurable via the `wait-exponential-backoff` ConfigMap and controlled by the `enable-wait-exponential-backoff` setting:
+
[source,yaml]
----
apiVersion: v1
kind: ConfigMap
metadata:
  name: feature-flags
  namespace: tekton-pipelines
data:
  enable-wait-exponential-backoff: "true"
#...
----

> * Before this update, the controller used a fixed thread count, limiting concurrency. With this update, you can override `threads-per-controller` by setting the `THREADS_PER_CONTROLLER` environment variable, allowing finer control over controller concurrency.

Reviewer 5: Please include the correct component controller to which this fix was done? Also is there any max limit?

* Before this update, upgrading Tekton could prevent updates to metadata, such as finalizers, on a completed `PipelineRun` or `TaskRun` due to specification drift. With this update, you can  update metadata on completed PipelineRuns and TaskRuns immediately following upgrades.

* Before this update, the OpenShift git resolver did not mount the trusted CA ConfigMap into the component system CA store, potentially causing certificate verification issues. With this update, the ConfigMap is mounted correctly, ensuring secure git operations.

* Before this update, the `git-clone` task failed with a `No such remote 'origin'` error messgae if the `origin` remote was missing from the repository. With this update, the task automatically adds the `origin` remote to the repository configuration, ensuring correct setup and successful cloning.

* Before this update, {pac} failed immediately when resource quotas were exceeded, canceling the run and interrupting user workflows. With this update, {pac} controllers retry and automatically rerun if resources become available, reducing unnecessary cancellations and improving pipeline reliability.

.Pruner

> * Before this update, the event-based pruner failed to process `PipelineRun` objects when the `historyLimit` setting was set to `0`, leaving them in a running state. With this update, the pruner handles `PipelineRuns` with `historyLimit` set to `0`, improving resource management.  

Reviewer 5: I am sorry, is this a release bug? I dont remember where this is coming from?

> * Before this update, the pruner could panic during upgrades if a setting was `nil`. With this update, the pruner handles `nil` values correctly and avoids crashes.

Reviewer 5: I think we could avoid this since this faced by non openshift users

+
[source,yaml]
----
kind: Repository
spec:
  settings:
    github:
      comment_strategy: "disable_all"
#...
----

* Before this update, the `PipelineRun starting` comment did not include a link to the {OCP} Console, making access less convenient. With this update, the comment includes a direct link to the `PipelineRun`.

* Before this update, empty commits in Bitbucket push events were processed by {pac}, causing the controller to crash. With this update, the controller ignores empty commits in the payload, preventing crashes.

* Before this update, {pac} incorrectly annotated `PipelineRun` objects modified by external controllers and marked as `started`. With this update, `PipelineRun` receives proper annotations on status change, helping ensure accurate tracking.

* Before this update, auto-merge was blocked if an unauthorized user opened a pull request (PR) and the target branch did not contain a `.tekton` directory. {pac} created a pending check that remained indefinitely, even after a repository admin approved the CI run with the `/ok-to-test` GitOps comment. With this update, the check is updated correctly after approval, allowing auto-merge to proceed as expected.

.CLI

> * Before this update, CLI reading pod logs from already deleted pods could cause a panic, leading to application crashes. With this update, deleted pod scenarios are handled properly, preventing crashes when reading logs.

Reviewer 4: please change tkn to opc

> * Before this update, the CLI `PipelineRunPending` status was displayed with incorrect coloring, making it harder to distinguish from other statuses. With this update, the Pending status color is correct for improved visual clarity.

Reviewer 4: please change tkn to opc

> * Before this update, CLI following logs could trigger a deadlock, causing the application to hang. With this update, the deadlock scenario in the log following functionality is fixed, ensuring reliable log streaming.

Reviewer 4: please change tkn to opc

----

* Before this update, the controller used a fixed thread count, limiting concurrency. With this update, you can override `threads-per-controller` by setting the `THREADS_PER_CONTROLLER` environment variable, allowing finer control over controller concurrency.

* Before this update, `podTemplate` fields in `TaskRunSpec` CR did not support parameter substitution, limiting matrix and multi-arch patterns. With this update, the controller  substitutes parameters in `podTemplate` fields for all `TaskRuns` and `TaskRunSpecs`.

* Before this update, the `onError` block in `pipeline` v1beta1 did not support variables. With this update, the controller  resolves the `onError` variables, improving error handling flexibility.

* Before this update, git resolver shell-outs sometimes failed to inherit environment variables, breaking environment-driven authentication or configuration. With this update, the resolver passes the pod environment correctly to all git subprocesses.

* Before this update, git resolver deployments could leave zombie git processes running, consuming resources. With this update, resolvers  use a `tini`-based image and entrypoint to actively clean up subprocesses.

> * * Before this update, upgrading {pipelines-shortname} could cause errors when updating metadata, such as finalizers, on completed `PipelineRuns` or `TaskRuns` due to specification drift. With this update, the issue is fixed.

Reviewer 6: ```suggestion
* Before this update, upgrading {pipelines-shortname} could cause errors when updating metadata, such as finalizers, on completed `PipelineRuns` or `TaskRuns` due to specification drift. With this update, the issue is fixed.
```

+
[source,yaml]
----
kind: Repository
spec:
  settings:
    github:
      comment_strategy: "disable_all"
#...
----

* Before this update, the `PipelineRun starting` comment did not include a link to the {OCP} Console, making access less convenient. With this update, the comment includes a direct link to the `PipelineRun`.

* Before this update, empty commits in Bitbucket push events were processed by {pac}, causing the controller to crash. With this update, the controller ignores empty commits in the payload, preventing crashes.

* Before this update, {pac} incorrectly annotated `PipelineRun` objects modified by external controllers and marked as `started`. With this update, `PipelineRun` receives proper annotations on status change, helping ensure accurate tracking.

* Before this update, auto-merge was blocked if an unauthorized user opened a pull request (PR) and the target branch did not contain a `.tekton` directory. {pac} created a pending check that remained indefinitely, even after a repository admin approved the CI run with the `/ok-to-test` GitOps comment. With this update, the check is updated correctly after approval, allowing auto-merge to proceed as expected.

.CLI

> * Before this update, `tkn` CLI reading pod logs from already deleted pods could cause a panic, leading to application crashes. With this update, deleted pod scenarios are handled properly, preventing crashes when reading logs.

Reviewer 7: @divyansh42 Shouldn't the CLI reference pointed to opc ?

* {pipelines-shortname} no longer emits deprecated metrics, such as `pipelinerun_count` or `running_taskruns_count`. You must update your dashboards and alerts with the new metrics, such as `pipelinerun_total` or `running_pipelineruns`.
+
[options="header"]
|===
| Deprecated | New

| `pipelinerun_count`
| `pipelinerun_total`

| `running_pipelineruns_count`
| `running_pipelineruns`

| `running_pipelineruns_waiting_on_pipeline_resolution_count`
| `running_pipelineruns_waiting_on_pipeline_resolution`

| `running_pipelineruns_waiting_on_task_resolution_count`
| `running_pipelineruns_waiting_on_task_resolution`

| `taskrun_count`
| `taskrun_total`

| `running_taskruns_count`
| `running_taskruns`

| `running_taskruns_throttled_by_quota_count`
| `running_taskruns_throttled_by_quota`

| `running_taskruns_throttled_by_node_count`
| `running_taskruns_throttled_by_node`
|===

[id="pipelines-kn-issues-1-20-0_{context}"]
=== Known issues

> * Using `tkn pr logs` in the OpenShift namespace may fail with repeated `Failed to list objects from openshift namespace` errors for both admin and non-admin users. 

Reviewer 6: ```suggestion
* Using `opc pr logs` in the OpenShift namespace may fail with repeated `Failed to list objects from openshift namespace` errors for both admin and non-admin users. 
```

Optimization to skip processing of already stored PipelineRuns::
* {tekton-results} skips processing of already stored `PipelineRuns`, improving performance and reducing API server calls. This optimization enhances overall efficiency and responsiveness, aiming to reduce server load and conserve resources.

.{pac}

Support for JSON body in incoming webhooks::
* {pac} supports passing incoming webhook parameters in the JSON POST body, while maintaining compatibility with URL query parameters:
+
[source, terminal]
----
curl -H "Content-Type: application/json" -X POST \
"https://control.pac.url/incoming" -d '{"repository":"repo","branch":"main", \
"pipelinerun":"target-pipelinerun","secret":"very-secure-shared-secret"}'
----
+
Using the request body for sensitive information reduces exposure in logs and improves security. For more details, see link:https://pipelinesascode.com/docs/guide/incoming_webhook/#recommended-post-json-body-method[Recommended (POST JSON body) method].

Detailed logging for GitHub API calls::

> * {pac} supports detailed logging for GitHub API calls, providing insights into API interactions, durations, and rate-limiting. By setting the controller log level to 'debug', you can troubleshoot complex issues more efficiently. This enhancement refactors code to instrument GitHub calls and affects all types of GitHub API calls within the provider. For more details, see link:https://pipelinesascode.com/docs/install/logging/#debugging-api-interactions[Debugging API interactions].

Reviewer 2: @ochromy can you refer upstream docs in downstream docs? e.g. https://pipelinesascode.com

|===

[id="pipelines-kn-issues-1-20-0_{context}"]
=== Known issues

.Pruner
* After upgrading from {pipelines-title} 1.19 to 1.20, the `tekton-pruner-default-spec`  config map values are overridden with default values. As a mitigation, maintain a copy of the `tekton-pruner-default-spec`  config map before upgrade and apply the same to the `TektonConfig` fields post upgrade. This issue affects only upgrade path and does not occur on fresh installations.

.CLI
* Using `opc pr logs` in the OpenShift namespace may fail with repeated `Failed to list objects from openshift namespace` errors for both admin and non-admin users. 

.{tekton-cache}
On IBM P environments, the `cache-fetch` step may fail with the error `failed to change ownership: operation not permitted`. This typically occurs because of filesystem permission restrictions on the underlying storage.

.{tekton-chains}
* Pod anti-affinity rules are not applied to `tekton-chains-controller` replicas.

.{tekton-hub}
* The `git-clone` task downloaded from {tekton-hub} displays version `0.9` instead of `0.10`. This occurs because the system sorts version strings lexicographically rather than using semantic versioning.

.{pac}

> * When upgrading to 1.20, the `hub_catalog_name` variable value is set to `tekton` by default and remains set to `tekton` after the upgrade. As a result, the system does not fetch the `git-clone` task from the correct catalog on {artifact-hub}.

Reviewer 2: but I still think that this shouldn't be included

* Before this update, git resolver deployments could leave zombie git processes running, consuming resources. With this update, resolvers  use a `tini`-based image and entrypoint to actively clean up subprocesses.

* Before this update, upgrading {pipelines-shortname} could cause errors when updating metadata, such as finalizers, on completed `PipelineRuns` or `TaskRuns` due to specification drift. With this update, the issue is fixed.

* Before this update, the OpenShift git resolver did not mount the trusted CA  config map into the component system CA store, potentially causing certificate verification issues. With this update, the  config map is mounted correctly, ensuring secure git operations.

* Before this update, the `git-clone` task failed with a `No such remote 'origin'` error messgae if the `origin` remote was missing from the repository. With this update, the task automatically adds the `origin` remote to the repository configuration, ensuring correct setup and successful cloning.

> * Before this update, {pac} failed immediately when resource quotas were exceeded, canceling the run and interrupting user workflows. With this update, {pac} controllers retry and automatically rerun if resources become available, reducing unnecessary cancellations and improving pipeline reliability.

Author: https://issues.redhat.com/browse/SRVKP-4165 - "PaC failing" in the description

> * Before this update, {pac} failed immediately when resource quotas were exceeded, canceling the run and interrupting user workflows. With this update, {pac} controllers retry and automatically rerun if resources become available, reducing unnecessary cancellations and improving pipeline reliability.

Author: But it is not in the `.{pac}` section og the fixed issues. Should I remove the mention of PAC specifically?

> * Before this update, {pac} failed immediately when resource quotas were exceeded, canceling the run and interrupting user workflows. With this update, {pac} controllers retry and automatically rerun if resources become available, reducing unnecessary cancellations and improving pipeline reliability.

Reviewer 2: You should move it to pipeline

| `running_pipelineruns_waiting_on_pipeline_resolution_count`
| `running_pipelineruns_waiting_on_pipeline_resolution`

| `running_pipelineruns_waiting_on_task_resolution_count`
| `running_pipelineruns_waiting_on_task_resolution`

| `taskrun_count`
| `taskrun_total`

| `running_taskruns_count`
| `running_taskruns`

| `running_taskruns_throttled_by_quota_count`
| `running_taskruns_throttled_by_quota`

| `running_taskruns_throttled_by_node_count`
| `running_taskruns_throttled_by_node`
|===

.{pac}

* {pac} automatically moves from using the deprecated {tekton-hub} to {artifact-hub} for pipeline and task resolution. After the upgrade some task references using short version pins, such as `0.2`, will not resolve correctly. You must update these pins to the full semantic version format, such as `0.2.0`, to ensure correct task resolution.

* The `hub_catalog_name` variable value remains set to `tekton` after upgrading to 1.20. As a result, the system does not fetch the `git-clone` task from the correct catalog on {artifact-hub}. After upgrading to 1.20, you must remove the variable from the {pac} config map by running the following command:
+
[source, terminal]
----

>  `cmd - oc patch configmap pipelines-as-code -n openshift-pipelines --type=json -p='[{"op": "remove", "path": "/data/hub-catalog-name"}]'`

Reviewer 6: ```suggestion
 oc patch configmap pipelines-as-code -n openshift-pipelines --type=json -p='[{"op": "remove", "path": "/data/hub-catalog-name"}]'
```

[id="new-features-1-20-0_{context}"]
== New features

In addition to fixes and stability improvements, the following sections highlight what is new in {pipelines-title} 1.20:

Support for running in FIPS-enabled environments::
* {pipelines-title} is designed for environments running in Federal Information Processing Standards (FIPS) mode. For more information about enabling {OCP} FIPS support, see the link:https://docs.redhat.com/en/documentation/openshift_container_platform/latest/html/installation_overview/installing-fips[{OCP} Support for FIPS cryptography].

Pod anti-affinity rule added to controller replicas::
* {pipelines-shortname} adds and applies the `preferredDuringSchedulingIgnoredDuringExecution` pod anti-affinity rule to controller replicas, except for the {tekton-chains} controllers. In High Availability (HA) setups, this rule distributes replicas across different nodes instead of scheduling them on the same node, improving resiliency, load balancing, and availability. No additional configuration is required.

New buildah-ns task for improved container build security::

> * {pipelines-shortname} adds the `buildah-ns` task. It improves container build security with user namespace isolation and remains compatible with the existing `buildah` task. For more details, see "Additional resources".

Reviewer 8: For more details, see "Additional resources". -> For more information, see the <xyz> section in Additional Resources.

Author: [lead-in sentences](https://github.com/openshift/openshift-docs/blob/main/contributing_to_docs/doc_guidelines.adoc#lead-in-link-sentences ) but I can change _details_ to _information_

readOnlyRootFilesystem enabled for default deployments::
* Containers in the {pipelines-shortname}, {tekton-results}, {tekton-chains}, and Manual Approval Gate deployments have `readOnlyRootFilesystem` setting enabled by default, improving security and compliance.

Tasks display friendly names in the web console::

> * Tasks in the {OCP} web console now use the displayName property instead of technical task names, improving readability and making tasks easier to search within the {pipelines-title} interface.

Reviewer 8: displayName -> `displayName`

{pipelines-shortname} on single-node OpenShift::
* Tech preview support for running {pipelines-shortname} on Single-Node OpenShift (SNO) is now available. This support is not intended for production deployments. {pipelines-shortname} on SNO has no additional limitations beyond the inherent constraints of a single-node cluster: limited scalability, no redundancy, and constrained concurrency.
For a cluster on libvirt UPI, the minimum hardware requirements are:

** 12 vCPUs
** 64 GB RAM
** 240 GB disk space

:FeatureName: {pipelines-shortname} on Single-Node OpenShift (SNO)
include::snippets/technology-preview.adoc[]

.Operator

Independent control of RBAC and CA bundle config map creation::
* With this update, you can independently control whether the {pipelines-shortname} Operator creates Role-Based Access Control (RBAC) resources and Trusted Certificate Authority (CA) bundle config maps in your cluster. This enhancement gives you more flexibility to fit your environment and helps avoid resource duplication. By default, both options are enabled to maintain the existing behavior. For more details, see "Additional resources".

.Pruner

> Event-based pruner configurable in TektonConfig CR::

Reviewer 8: TektonConfig -> `TektonConfig`?

> * You can enable and configure the event-based `tektonpruner` directly in the `TektonConfig` custom resource (CR). Observability enhancement is also included with the new metrics specific to pruner. 

Reviewer 8: You can enable and configure the event-based tektonpruner in the TektonConfig custom resource (CR). This update also adds observability enhancements with new pruner-specific metrics.

:FeatureName: The event-based pruner
include::snippets/technology-preview.adoc[]

.{tekton-triggers}

Optional installation of {tekton-triggers} through the Operator::

> * You can choose to not install {tekton-triggers} when deploying {pipelines-shortname} through the Operator. This enhancement provides finer control over component installation, allowing flexibility for environments where triggers are managed separately. The default setting is `false`.

Reviewer 8: When deploying {pipelines-shortname} through the Operator, you can choose not to install {tekton-triggers}. This update provides finer control over component installation and supports environments where triggers are managed independently. The default value is `false`.

+
[source, yaml]
----
apiVersion: operator.tekton.dev/v1alpha1
kind: TektonConfig
metadata:
  name: config
spec:
  trigger:
    disabled: true
#...
----

.{tekton-results}

New flag to disable live collection in tekton-results-watcher::
* {tekton-results} supports a new flag for the `tekton-results-watcher` controller to disable live collection of runs by setting the following flag in the `TektonConfig` CRD:
+

> [source, yaml]

Reviewer 8: Same as above for a possible addition of a lead-in sentence.

Optimization to skip processing of already stored PipelineRuns::
* {tekton-results} skips processing of already stored `PipelineRuns`, improving performance and reducing API server calls. This optimization enhances overall efficiency and responsiveness, aiming to reduce server load and conserve resources.

.{pac}

Support for JSON body in incoming webhooks::
* {pac} supports passing incoming webhook parameters in the JSON POST body, while maintaining compatibility with URL query parameters:
+
[source, terminal]
----
curl -H "Content-Type: application/json" -X POST \
"https://control.pac.url/incoming" -d '{"repository":"repo","branch":"main", \
"pipelinerun":"target-pipelinerun","secret":"very-secure-shared-secret"}'
----
+
Using the request body for sensitive information reduces exposure in logs and improves security. For more details, see link:https://pipelinesascode.com/docs/guide/incoming_webhook/#recommended-post-json-body-method[Recommended (POST JSON body) method].

Detailed logging for GitHub API calls::

> * {pac} supports detailed logging for GitHub API calls, providing insights into API interactions, durations, and rate-limiting. By setting the controller log level to 'debug', you can troubleshoot complex issues more efficiently. This enhancement refactors code to instrument GitHub calls and affects all types of GitHub API calls within the provider. For more details, see link:https://pipelinesascode.com/docs/install/logging/#debugging-api-interactions[Debugging API interactions].

Reviewer 8: For more details -> For more information.

New auto-configure-repo-repository-template option::
* {pac} adds the `auto-configure-repo-repository-template` setting, a new option in `auto-configure-new-github-repo` feature setting to consolidate repository creation within a single namespace, streamlining management for those with multiple repositories. For more details, see "Additional resources".

Support for relative task references in remote Pipeline definitions::

> * You can use relative paths to reference tasks from within a remote `Pipeline` definition. The {pac} resolver automatically builds the full URL for the task based on the location of the remote `PipelineSpec` definition. This improves pipeline organization and portability by keeping tasks and pipelines co-located, which is especially useful when working with tagged releases or branches. For more details, see link:https://pipelinesascode.com/docs/guide/resolver/#overriding-tasks-from-a-remote-pipeline-on-a-pipelinerun[Overriding tasks from a remote pipeline on a PipelineRun].

Reviewer 8: You can now use relative paths to reference tasks within a remote `Pipeline` definition. The {pac} resolver automatically constructs/builds the entire/complete URL for the task based on the location of the remote `PipelineSpec` definition. This enhancement improves pipeline organization and portability, particularly when working with tagged releases or branches. For more information, see link:https://pipelinesascode.com/docs/guide/resolver/#overriding-tasks-from-a-remote-pipeline-on-a-pipelinerun[Overriding tasks from a remote pipeline on a PipelineRun].

:FeatureName: Overriding tasks from a remote Pipeline
include::snippets/technology-preview.adoc[]

New dynamic pull_request_number variable for push events::

> * Pull requests that trigger a push event automatically populate a new {pac} dynamic `pull_request_number` variable, allowing for a clear reference to the specific pull request associated with the push event and improving traceability. For more details, see "Additional resources".

Reviewer 2: ```suggestion
* Pull requests that trigger a push event automatically populates a new {pac} dynamic `pull_request_number` variable, allowing for a clear reference to the specific pull request associated with the push event and improving traceability. For more details, see "Additional resources".
```

[id="pipelines-breaking-changes-1-20-0_{context}"]
=== Breaking changes

> * The public instance of Tekton Hub (`hub.tekton.dev`) is deprecated and will be removed in a future release. For more details, see "Additional resources".

Reviewer 8: Check earlier feedback for more details, see "Additional Resources".

* Before this update, the {OCP} Console stated a 'Cancelling' state for `PipelineRun` objects with failed `finally` tasks, even when the actual status of the `PipelineRun` was `Canceled`, causing confusion. With this update, the issue is fixed.

* Before this update, the `PipelineRun` Events tab did not show events, due to an issue with the Pipeline Console plugin. As a consequence, you could not monitor the events in real-time. With this update, the issue is fixed.

* Before this update, the pipeline builder fetched only a limited number of tasks from {artifact-hub}, resulting in some tasks not being available. This resulted in an incomplete pipeline creation. With this update, the interface fetches the complete list of tasks from {artifact-hub} on search, showing all available tasks in the pipeline builder.

> * Before this update, running `fbc-fips-check-oci-ta` tasks during git cloning caused temporary resource unavailability, stalling the git resolver and preventing the resolution of `ResolutionRequests`. This resulted in a build failure with the following error message: `cannot fork() for remote-https: Resource temporarily unavailable`. With this update, the issue is fixed.

Reviewer 9: This task is konflux specific isn't it ? For a customer that doesn't know anything about konflux, this entry is a bit confusing imo.

Author: Should we rewrite it as:
```suggestion
* Before this update, running the Konflux-specific `fbc-fips-check-oci-ta` tasks during git cloning caused temporary resource unavailability, stalling the git resolver and preventing the resolution of `ResolutionRequests`. This resulted in a build failure with the following error message: `cannot fork() for remote-https: Resource temporarily unavailable`. With this update, the issue is fixed.
```

* Before this update, `TaskRun` and `CustomRun` creation could fail immediately on transient mutating-webhook timeouts, causing flakiness on busy clusters. With this update, the system  applies exponential backoff, configurable via the `wait-exponential-backoff`  config map and controlled by the `enable-wait-exponential-backoff` setting:
+
[source,yaml]
----
apiVersion: v1
kind: ConfigMap
metadata:
  name: feature-flags
  namespace: tekton-pipelines
data:
  enable-wait-exponential-backoff: "true"
#...
----

> * Before this update, the controllers used a fixed thread count, limiting concurrency. With this update, you can override `threads-per-controller` by setting the `THREADS_PER_CONTROLLER` environment variable, allowing finer control over controller concurrency.

Reviewer 9: Is it on a specific controller ? I thought tektoncd/pipeline supported this for a long time already.

* Before this update, `podTemplate` fields in `TaskRunSpec` CR did not support parameter substitution, limiting matrix and multi-arch patterns. With this update, the controller substitutes parameters in `podTemplate` fields for all `TaskRuns` and `TaskRunSpecs`.

* Before this update, the `onError` block in `pipeline` v1beta1 did not support variables. With this update, the controller  resolves the `onError` variables, improving error handling flexibility.

* Before this update, git resolver shell-outs sometimes failed to inherit environment variables, breaking environment-driven authentication or configuration. With this update, the resolver passes the pod environment correctly to all git subprocesses.

> * Before this update, git resolver deployments could leave zombie git processes running, consuming resources. With this update, resolvers  use a `tini`-based image and entrypoint to actively clean up subprocesses.

Reviewer 9: @athorp96 isn't `tini` based only upstream ?

Reviewer 10: `tini` is [used downstream](https://github.com/openshift-pipelines/tektoncd-pipeline/blob/60a1d516229066f70233c56392a66cbc6a16e618/.konflux/dockerfiles/resolvers.Dockerfile#L54) as well, but it's not a `tini`-based image. It's not new to 1.20 though. Both of these changes were introduced in 1.19: [pod environment in git](https://github.com/openshift-pipelines/tektoncd-pipeline/blob/0a7000aba8ab139c014af62b191890818ca9ab82/upstream/pkg/resolution/resolver/git/repository.go#L122) and [tini entrypoint](https://github.com/openshift-pipelines/tektoncd-pipeline/blob/0a7000aba8ab139c014af62b191890818ca9ab82/.konflux/dockerfiles/resolvers.Dockerfile#L53)
